aspnetcore/fundamentals/configuration/options/samples/6.x/OptionsValidationSample/Program.cs

```diff
@@ -0,0 +1,190 @@
+#define P2   // DEFAULT MC XM P1  P2
+#if NEVER
+#elif DEFAULT
+#region snippet
+using OptionsValidationSample.Configuration;
+
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.AddControllersWithViews();
+
+builder.Services.AddOptions<MyConfigOptions>()
+            .Bind(builder.Configuration.GetSection(MyConfigOptions.MyConfig))
+            .ValidateDataAnnotations();
+
+var app = builder.Build();
+#endregion
+
+if (!app.Environment.IsDevelopment())
+{
+    app.UseExceptionHandler("/Home/Error");
+    app.UseHsts();
+}
+
+app.UseHttpsRedirection();
+app.UseStaticFiles();
+
+app.UseRouting();
+
+app.UseAuthorization();
+
+app.MapControllerRoute(
+    name: "default",
+    pattern: "{controller=Home}/{action=Index}/{id?}");
+
+app.Run();
+#elif MC
+#region snippet_mc
+using OptionsValidationSample.Configuration;
+
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.AddControllersWithViews();
+
+builder.Services.AddOptions<MyConfigOptions>()
+            .Bind(builder.Configuration.GetSection(MyConfigOptions.MyConfig))
+            .ValidateDataAnnotations()
+        .Validate(config =>
+        {
+            if (config.Key2 != 0)
+            {
+                return config.Key3 > config.Key2;
+            }
+
+            return true;
+        }, "Key3 must be > than Key2.");   // Failure message.
+
+var app = builder.Build();
+#endregion
+
+if (!app.Environment.IsDevelopment())
+{
+    app.UseExceptionHandler("/Home/Error");
+    app.UseHsts();
+}
+
+app.UseHttpsRedirection();
+app.UseStaticFiles();
+
+app.UseRouting();
+
+app.UseAuthorization();
+
+app.MapControllerRoute(
+    name: "default",
+    pattern: "{controller=Home}/{action=Index}/{id?}");
+
+app.Run();
+#elif XM
+#region snippet_xm
+using Microsoft.Extensions.Options;
+using OptionsValidationSample.Configuration;
+
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.AddControllersWithViews();
+
+builder.Services.Configure<MyConfigOptions>(builder.Configuration.GetSection(
+                                        MyConfigOptions.MyConfig));
+
+builder.Services.AddSingleton<IValidateOptions
```

Comment:
Is this the correct replacement for the following?
```
public void ConfigureServices(IServiceCollection services)
{
    services.Configure<MyConfigOptions>(Configuration.GetSection(
                                        MyConfigOptions.MyConfig));
    services.TryAddEnumerable(ServiceDescriptor.Singleton<IValidateOptions
                              <MyConfigOptions>, MyConfigValidation>());
    services.AddControllersWithViews();
}
``

Reply:
@halter73 See [V6 IValidateOptions for complex validation](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-6.0&branch=pr-en-us-23750#ivalidateoptions-for-complex-validation) and Previous [V5 IValidateOptions for complex validation](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-6.0#ivalidateoptions-for-complex-validation) 

It might be easier to review the code from the links than from here.

Is this the correct replacement for the following?
```
public void ConfigureServices(IServiceCollection services)
{
    services.Configure<MyConfigOptions>(Configuration.GetSection(
                                        MyConfigOptions.MyConfig));
    services.TryAddEnumerable(ServiceDescriptor.Singleton<IValidateOptions
                              <MyConfigOptions>, MyConfigValidation>());
    services.AddControllersWithViews();
}
```